[GPII-3868]: Fixes for Stackdriver Trace

natarajaya opened the pull request:
This solution fixes Stackdriver Trace by providing SA credentials for `cloud-trace-nodejs` plugin for `universal`.
It does not require legacy metadata API to retrieve tokens, but stores SA credentials as K8s secret to mount into GPII containers instead.

There are couple things that can be improved here:
* SA key rotation – does not seem like TF's [google_service_account_key](https://www.terraform.io/docs/providers/google/r/google_service_account_key.html) resource is capable of this. There are couple validity attributes, but that functionality is unclear and needs to be tested.
* Duplicating Istio's VirtualService and ServiceEntry resources for egress traffic in every chart. But the only way to apply them outside of Helm I can think of is with raw `kubectl` commands, so I don't think thats better.

stepanstipl commented:
I think this looks quite good, the part I'm not happy about is static credentials being generated and stored across various places in the system.

What about a slight variation on this: as we have a mechanism to assign individual GCP service accounts to pods (via annotations), it should be possible to only generate the `/etc/credentials/credentials.json` on the fly, for example with init container. Once the pod would start, init container would use `gcloud iam service-accounts keys create` to generate a new key and save it to in-memory volume.

natarajaya commented:
@stepanstipl I have the same concern about static credentials, however, I can think of relatively low-cost solution to seamlessly rotate them with TF's `null_resource` and `gcloud`.

> Once the pod would start, init container would use gcloud iam service-accounts keys create to generate a new key

But `gcloud` itself needs credentials to run – am I missing something here?

stepanstipl commented:
> But gcloud itself needs credentials to run – am I missing something here?

But `gcloud` can handle metadata correctly. So once you add the `traffic.sidecar.istio.io/excludeOutboundIPRanges` annotation to allow traffic to metadata, `gcloud` has credentials available:

```
$ gcloud auth list
                           Credentialed Accounts
ACTIVE  ACCOUNT
*       gke-cluster-pod-default@gpii-gcp-dev-stepan.iam.gserviceaccount.com

To set the active account, run:
    $ gcloud config set account `ACCOUNT`

$ gcloud projects list
PROJECT_ID           NAME                 PROJECT_NUMBER
gpii-gcp-dev-stepan  gpii-gcp-dev-stepan  374774757640
```

And we even have the mechanism to assign specific service account to pods - via the `accounts.google.com/service-account` annotation.

-> So no static credentials.

So perhaps creating dedicated svc. an account with extra permissions to generate keys for itslef might work.

stepanstipl commented on shared/charts/gpii-flowmanager/templates/virtual_service_cloudtrace.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.enableStackdriverTrace }}
```
`VirtualService` is not needed to allow access to external service, only the `ServiceEntry` -> remove

natarajaya replied:
Documentation says otherwise:
https://archive.istio.io/v1.0/docs/tasks/traffic-management/egress/#configuring-the-external-services

```Note that for TLS protocols, including HTTPS, the TLS VirtualService is required in addition to the ServiceEntry.```

natarajaya replied:
I double checked and everything seems to be working without `VirtualService` entries – outdated documentation?

stepanstipl replied:
Yeah, that's my experience. `VirtualService` is typically required only if you want to affect how routing works (ie. split between 2 services or drop traffic or smth. like that).

Also, you're referencing some old version of docs (I know we're using 1.0), in current docs it's not there anymore: https://istio.io/docs/tasks/traffic-management/egress/egress-control/#access-an-external-https-service

natarajaya replied:
So it turns out that `VirtualService` is required after all (#448) :)

stepanstipl commented on shared/charts/gpii-flowmanager/templates/virtual_service_googleapis.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.enableStackdriverTrace }}
```
`VirtualService` is not needed to allow access to external service, only the `ServiceEntry` -> remove

natarajaya replied:
Same here.

stepanstipl commented on shared/charts/gpii-flowmanager/templates/service_entry_googleapis.yaml:
```diff
@@ -0,0 +1,16 @@
+  name: {{ template "flowmanager.name" . }}-googleapis
+spec:
+  hosts:
+  - www.googleapis.com
```
Why do we need access to `www.googleapis.com`? Also would it make sense to merge this in one `ServiceEntry` such as:
```yaml
  hosts:
  - www.googleapis.com
  - cloudtrace.googleapis.com
```
?

natarajaya replied:
Looks like when SA credentials provided, access to `www.googleapis.com` is not needed anymore.

stepanstipl commented on shared/charts/gpii-preferences/templates/deployment.yaml:
```diff
@@ -9,7 +9,6 @@ spec:
     rollingUpdate:
       maxSurge: {{ .Values.rollingUpdate.maxSurge }}
       maxUnavailable: {{ .Values.rollingUpdate.maxUnavailable }}
-    type: RollingUpdate
```
Why is this removed? Probably doesn't make any difference as `RollingUpdate` is the default... but?

natarajaya replied:
This is exactly why it is removed :) It is default type and it was present in preferences chart and missing in flowmanager chart, so I removed it for consistency. I could add it into flowmanager chart if you prefer that.

stepanstipl replied:
Ok, thanks for the explanation, I think it's fine. Consistency is good 👍

stepanstipl commented on shared/charts/gpii-preferences/templates/service_entry_googleapis.yaml:
```diff
@@ -0,0 +1,16 @@
+  name: {{ template "preferences.name" . }}-googleapis
+spec:
+  hosts:
+  - www.googleapis.com
```
Same as above, why is `www.googleapis.com` needed & might make sense to merge in one `SE`.

stepanstipl commented on shared/charts/gpii-preferences/templates/virtual_service_cloudtrace.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.enableStackdriverTrace }}
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
```
Same, `VS` is not needed to allow traffic.

stepanstipl commented on shared/charts/gpii-preferences/templates/virtual_service_googleapis.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.enableStackdriverTrace }}
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
```
Same ^.

stepanstipl commented on shared/charts/gpii-preferences/values.yaml:
```diff
@@ -10,7 +10,11 @@ datasourceListenPort: 5984
 nodeEnv: gpii.config.preferencesServer.standalone.production
+
 enableStackdriverTrace: false
+## When Stackdriver Trace is enabled SA credentials must be provided in a secret,
```
Why double `#` comments?

natarajaya replied:
Double `#` helps to distinguish between actual comments and commented configuration blocks. It follows Helm style convention:
https://github.com/gpii-ops/gpii-infra/blob/master/shared/charts/couchdb/values.yaml
https://github.com/gpii-ops/gpii-infra/blob/master/shared/charts/nginx-ingress/values.yaml

stepanstipl replied:
This seems pretty weird to me, I would still prefer single comments, but I can live with that :)

stepanstipl commented on shared/charts/gpii-flowmanager/values.yaml:
```diff
@@ -11,7 +11,11 @@ datasourceHostname: "http://admin:password@couchdb-svc-couchdb.gpii.svc.cluster.
 preferencesUrl: http://preferences.gpii.svc.cluster.local
+
 enableStackdriverTrace: false
+## When Stackdriver Trace is enabled SA credentials must be provided in a secret,
```
Double `#`

natarajaya replied:
Same here.

natarajaya commented:
> And we even have the mechanism to assign specific service account to pods - via the accounts.google.com/service-account annotation. So no static credentials.

@stepanstipl But there would still be static credentials, your proposal just handles them on K8s level instead of TF level.

> I would still propose creating dedicated GCP svc accounts for each of the workloads (preferences & flowmanager), then the code managing the service keys could live in corresponding gpii-* module and module gpii-istio can stay as it is

Agreed and agreed. Another complication of renaming module that contains K8s resources – TF not always can handle this gracefully. In this case TF would try to recreate `kubernetes_namespace` declared in `gpii-istio` and fail unless it is manually imported into new module's TF state. So I am more than happy to leave `gpii-istio` as it is.

mrtyler commented on gcp/modules/gpii-flowmanager/sa.tf:
```diff
@@ -0,0 +1,23 @@
+variable "sa_id" {
```
I think this will be clearer if you spell out "serviceaccount" (or "service_account", whichever is more consistent) here and in the filename. (The abbreviation "sa" is fine in places like terraform resource names, where the "service_account" part is included elsewhere. You could also use Stepan's "this" convention.)

stepanstipl commented:
OK, thanks @natarajaya. Looks good to me 👍 

Please try to avoid using "force-push" once the code is in PR and has been reviewed -> this breaks most of the GitHub review functionality (like view changes since the last review) and makes it much harder for the reviewers to track the evolution of code.

amatas commented:
LGTM, 
I reviewed your comments and the log of the pad of yesterday. How can we track Google's fixing of the tracing library, a ticket? a comment in this code?

natarajaya commented:
Thanks for all the reviews!

* This is ready to be merged as a straightforward solution to get tracing working while metadata issue is still affecting GKE. 
* I tried Stepan's proposal to handle SA credentials on K8s level, but since `trace-agent` is incapable of handling credentials rotation on the fly, this does not seem to be feasible, so we don't have an alternative atm.
* I think we can keep this long-term (worst case scenario in case it takes a long time for Google to address metadata issues) if we implement a way to rotate SA credentials used by deployments (proposed in #435).

> Please try to avoid using "force-push" once the code is in PR and has been reviewed -> this breaks most of the GitHub review functionality (like view changes since the last review)

@stepanstipl No worries, I am well aware of that – accidentally pushed bunch of commits into wrong branch late at night. I don't think it affects this PR in any way since it is ready to be merged.

> How can we track Google's fixing of the tracing library, a ticket? a comment in this code?

@amatas It is in the ticket and in the PR with metadata issues (#430).

stepanstipl commented:
> - I tried Stepan's proposal to handle SA credentials on K8s level, but since trace-agent is incapable of handling credentials rotation on the fly, this does not seem to be feasible, so we don't have an alternative atm.
> - I think we can keep this long-term (worst case scenario in case it takes a long time for Google to address metadata issues) if we implement a way to rotate SA credentials used by deployments (proposed in #435).

So these are two separate issues:
- Where to handle the issuance of the credentials - in or outside of the kube cluster
- And the credentials rotation

As it turned out the predicate our previous conversation was based on, that the app will notice the change of credentials without restart, is not true. This means that should we want to rotate the credentials we have to restart the pods - yet this can still be done in or outside of the cluster (more about that in #435 ).

stepanstipl commented:
@natarajaya one thing that occurred to me, as you tracked this down to be http header issue - have you tried how this behaves with the new GKE Workload Identity? https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity

natarajaya commented:
@stepanstipl This duplicates conversation from #435, let's keep it there.

stepanstipl commented:
I did try to enable GKE Workload Identity on 1.12 cluster and tracing seems to work out of the box. Please see #435 for details and my proposal on how to go about this.

natarajaya commented:
@stepanstipl I was able to finally track down the real reason of metadata issues to `service-account-assigner`! More details [here](https://github.com/gpii-ops/gpii-infra/pull/435#issuecomment-508883780).

natarajaya commented:
Update: I was able to fix `service-account-assigner` to always add `Metadata-Flavor` header to its responses and now have trace working in my dev cluster without SA credentials!

natarajaya commented:
Ok, so what remains from the changes proposed in this PR:
* Istio service entries to allow egress to trace API for flowmanager and preferences.
* Annotations for flowmanager and preferences deployments to enable off-mesh egress to metadata.
* Reformatted flowmanager and preferences chart READMEs.

Maybe I should've cherry-picked them to new branch and opened separate PR? But at the same time they are not hard to review from scratch and history may still be useful...

What is still missing:
* Versions file needs to be updated with new `service-account-assigner` tag once https://github.com/gpii-ops/k8s-gke-service-account-assigner/pull/2 and `docker-image-service-account-assigner` PRs are merged.

stepanstipl commented:
So currently there seems to be issue with the `ServiceEntry` breaking egress traffic - see https://issues.gpii.net/browse/GPII-4015 for details.

Otherwise looks good, pity you didn't keep the dedicated SVC accounts around at least on Kube level - I thought that was still a good imrpovement.

natarajaya commented:
@stepanstipl With VirtualServices added to preferences and flowmanager charts, I am still able to reproduce Locust issues from GPII-4015.

However, it took me 4 consecutive `rake destroy && rake && rake test_flowmanager` runs, and on 5th run tests went smoothly again.

So I propose to merge this as it is, in case we see Locust issues in CI – we have plan B, upgrade to 1.13 with Istio 1.1 (#450).

stepanstipl commented:
> So I propose to merge this as it is, in case we see Locust issues in CI – we have plan B, upgrade to 1.13 with Istio 1.1 (#450).

Thanks @natarajaya. Sounds good. I agree with the plan, the VS definitions look correct.
